### PR TITLE
Add mutex protection on ConversationHandler

### PR DESCRIPTION
**Pull request by SnowyCoder**
Advances #1029 adding thread safety on the mutable variables of the class (`conversations` and `timeout_jobs`).
The original issue also asked to internalize all the other properties but it would incur breaking changes and a lot of refactoring, I'll leave this task to someone more proficient with the codebase.

### CODE REVIEW

**Comment by Poolitzer** [comment]
Hey man, this looks good to me, but there should be definitly someone else to look over this *cough cough* @eldin? *cough cough*. Can you remove the #Todo line though and either put it in this conversation or in the related issue? Then codacy should be fine with it.

File: telegram/ext/conversationhandler.py [hunk 1]
```diff
@@ -319,46 +324,62 @@ def handle_update(self, update, dispatcher, check_result, context=None):
...
 
-        if timeout_job is not None:
-            timeout_job.schedule_removal()
-        if self.conversation_timeout and new_state != self.END:
-            self.timeout_jobs[conversation_key] = dispatcher.job_queue.run_once(
-                self._trigger_timeout, self.conversation_timeout,
-                context=_ConversationTimeoutContext(conversation_key, update, dispatcher))
+        with self.timeout_jobs_lock:
+            # Remove the old timeout job (if present)
+            # TODO: should this be called before the child handle_update? the timeout could be called while the child
```
Comment (jh0ker) [inline telegram/ext/conversationhandler.py]: Yes, I believe you are correct, this would be more consistent.

**Comment by SnowyCoder** [comment]
I've addressed the todo, should I squash the commits?
Sorry for the trouble, I wanted to help you a bit but It's my first time doing a PR in a big repo :sweat_smile:.

**Comment by Poolitzer** [comment]
No need to, we will before merging into master. Just go ahead and make as many commits as you want :)

**Comment by jh0ker** [comment]
@SnowyCoder Now you have the chance to get the true _**big repo** experience™️_ 😉 
It would be great to have a test for that last change you made regarding the removal of the timeout job. Basically a test that demonstrates that the timeout job will be removed and not executed, even if you have a child handler that starts before the timeout would run out and ends after.

Refer to https://github.com/python-telegram-bot/python-telegram-bot/blob/master/tests/test_conversationhandler.py line 414 and following for inspiration.

File: telegram/ext/conversationhandler.py [hunk 2]
```diff
@@ -162,7 +163,9 @@ def __init__(self,
         Set by dispatcher"""
 
         self.timeout_jobs = dict()
+        self.timeout_jobs_lock = Lock()
         self.conversations = dict()
+        self.conversations_lock = Lock()
```
Comment (jh0ker) [inline telegram/ext/conversationhandler.py]: Please prefix those locks with a `_` so we're not committed to this API

**Comment by SnowyCoder** [comment]
I've been studying the test code and I noticed something strange: the assignment at line 494 (that have moved to 552 in the master branch) is duplicated, it seems like a merge error or a typo.

**Comment by jh0ker** [comment]
@SnowyCoder Interesting find ^^

**Comment by jh0ker** [comment]
Thanks a lot for your contribution @SnowyCoder!

**Comment by SnowyCoder** [comment]
Thank you for the support!

**Pull request merged**